Add before_register hook to RSpec World

bootstraponline opened the pull request:
Fix https://github.com/rspec/rspec-core/issues/2078

Use case: 
- [sauce_rspec/rspec.rb](https://github.com/bootstraponline/sauce_rspec/blob/db6904bfd8fc3b7038f77ad64960ebaef5171a0f/lib/sauce_rspec/rspec.rb)
- [SauceRSpec config](https://github.com/bootstraponline/sauce_connect_ruby/blob/f2b5e9cd0c941492ced15912b26cb36987e3439c/helper/sauce_helper.rb#L4)

Enables running an unmodified test suite on multiple Selenium browsers.

myronmarston commented on lib/rspec/core/world.rb:
```diff
@@ -19,6 +19,17 @@ def initialize(configuration=RSpec.configuration)
+      # Invokes block before registering an example group
+      def before_register(&block)
+        before_register_callbacks << block
+      end
```
I think this should go on config instead of world.  Config is where we let users configure callbacks.

Also, example group registration isn't really a concept we've exposed into our API up to now.  It's purely an internal.  I think that this feature would fit better with RSpec's terminology if it was something like `on_example_group_definition`, e.g.:

``` ruby
RSpec.configure do |c|
  c.on_example_group_definition do |example_group|
    # do something with the newly defined example group
  end
end
```

bootstraponline replied:
Makes sense. I'll fix.

myronmarston commented on spec/rspec/core/world_spec.rb:
```diff
@@ -21,6 +21,33 @@ module RSpec::Core
+            end
+          end
+
+          example_group.instance_variable_set(:@examples, new_examples)
```
Messing with the instance variables of the example group is definitely not something we'll support.  As such, I'd prefer the spec for this (which should show an example of the usage of this feature) not demonstrate it by doing something that we don't support users doing.

bootstraponline replied:
This is how the feature is used (duplicate the existing examples in an example group and change their id/metadata). I can rewrite the spec to just modify the metadata which will avoid the instance variable calls.

I'm assuming you don't want users to have the ability to redefine examples on an example group?

myronmarston replied:
If there's a need for users to mutate some of the state of the example group, we'll want to provide explicit APIs to do so.  Mutating instance variables will never be supported.

That said, for metadata, we already provide an API for mutating metadata:

http://rspec.info/documentation/3.2/rspec-core/RSpec/Core/Configuration.html#define_derived_metadata-instance_method

For the id, I don't think we can support users mutating it.  IDs are meant to be treated as opaque strings that RSpec generates such that future RSpec runs can use them to refer to the same ID.  They are parsed and used in certain ways such that allowing users to mutate them will almost certainly break features that use IDs.

Seeing how you intend to use this callback makes me think that we need to provide a different way to solve your problem.  I need to go read the example you put together in #2078 to understand what exactly you're trying to do.

bootstraponline replied:
An accessor method on example_group for examples would solve that issue.

The only reason I'm mutating IDs is RSpec will not recognize the newly created tests because they have the same id as other tests. These methods are created dynamically so location is not a unique identifier.

metadata is great. [I hit a limitation related to memoization](https://github.com/bootstraponline/sauce_rspec/blob/db6904bfd8fc3b7038f77ad64960ebaef5171a0f/lib/sauce_rspec/rspec.rb#L30) however and I had to monkey patch in an accessor method to bypass this. I updated the spec for this feature to use metadata so users can see a supported way of using the hook.



> Seeing how you intend to use this callback makes me think that we need to provide a different way to solve your problem.

I think this PR gets us partially the way there. I agree it's not a complete solution though.

bootstraponline commented:
I updated the PR to use `on_example_group_definition` instead of `before_register`. I also fixed the spec so it shows a supported use case. If there's a better way to accomplish the goal without this hook then that'd be great as well.

myronmarston commented:
@bootstraponline I think what you have here is reasonable as far as a way to hook in to example group creation in RSpec.  But to generate additional examples that are essentially clones of existing ones w/ some modifications, we should provide you with some new APIs.  Here's what I'm thinking...

* `ExampleGroup` needs new public APIs like `ExampleGroup.add_example` and `ExampleGroup.remove_example`
* `Example` needs a new public API like `duplicate_with(metadata_overrides)`.

Then for your case I think you could do something like this:

``` ruby
RSpec.configure do |config|
  config. on_example_group_definition do |group|
    group.examples.each do |example|
      group.remove_example example

      [{:browserName => :firefox}, {:browserName => :chrome}].each do |cap|
        group.add_example example.duplicate_with(:cap => cap)
      end
    end
  end
end
```

Thoughts?

bootstraponline commented:
That looks awesome!

myronmarston commented:
> That looks awesome!

Want to open up PRs adding those APIs? (and/or add them to this one)?

bootstraponline commented:
> Want to open up PRs adding those APIs? (and/or add them to this one)?

I think they're better as independent PRs. I'll work on submitting a new PR for `ExampleGroup` and another for `Example`. 

add_example/remove_example should be easy. duplicate_with is a bit more tricky since rspec will delete tests if they share the same id so I'll have to change it to something else (maybe add a hash of the metadata?).

myronmarston commented:
> duplicate_with is a bit more tricky since rspec will delete tests if they share the same id so I'll have to change it to something else (maybe add a hash of the metadata?).

I'd recommend not doing anything with the ids.  In fact, don't even use `dup` or `clone` -- instead, just instantiate a new instance of `Example` and RSpec will take care of the rest.  Essentially you'd do something similar to what RSpec does here:

https://github.com/rspec/rspec-core/blob/b8f0270f57253dbf02087b120824d84af6e35dce/lib/rspec/core/example_group.rb#L139-L145

In fact, it might be as simple as:

``` ruby
Example.new(example.example_group_class, example.description, example.metadata, example.example_block)
```

bootstraponline commented:
Ok, I'll try it.

bootstraponline commented:
I squashed to one commit and Travis is passing. appveyor had an internal server error unrelated to the PR.

bootstraponline commented:
I submitted https://github.com/rspec/rspec-core/pull/2095 for add/remove example. Working on duplicate_with next.

JonRowe commented:
LGTM
